File: wordSubsets/fastestSolution.cpp
// link: https://leetcode.com/problems/word-subsets/
class Solution {
public:        
        
    vector<string> wordSubsets(vector<string>& A, vector<string>& B) {
        int hash[26] = {};
        
        for (const auto& b : B) {
            int subHash[26] = {};
            for (char ch : b) {
                char idx = ch - 'a';
                subHash[idx]++;
                hash[idx] = max(subHash[idx], hash[idx]);
            }
        }
        
        vector<string> result;
        for (const auto& a : A) {
            int subHash[26] = {};
            for (char ch : a) {
                subHash[ch - 'a']++;
            }
            
            bool isUniversal = true;
            for (char i = 0; i < 26; i++) {
                if (hash[i] > subHash[i]) {
                    isUniversal = false;
                    break;
                }
            }
            
            if (isUniversal) {
                result.push_back(a);
            }
        }
        return result;
    }
};

File: word_pattern/fastest_solution.py
# https://leetcode.com/problems/word-pattern/
from typing import List


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        var: dict = {}
        sl: List[str] = s.split()
        pl: int = len(pattern)
        if pl == len(sl) and len(set(list(pattern))) == len(set(sl)):
            for p in range(pl):
                if pattern[p] not in var.keys():
                    var[pattern[p]] = sl[p]
                else:
                    if var[pattern[p]] != sl[p]:
                        return False
            return True
